streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -1378,4 +1374,168 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+            subtopologyId,
+            new HashSet<>(nodesByName.values())));
+    }
+
+    private HashMap<String, TopologyDescription.Node> addAllNodesWithoutPredecessorsOrSuccessors(final Set<String> nodeNames) {
+        final HashMap<String, TopologyDescription.Node> nodesByName = new HashMap<>();
+
+        for (final String nodeName : nodeNames) {
+            final NodeFactory nodeFactory = nodeFactories.get(nodeName);
+            final TopologyDescription.Node node;
+
+            if (nodeFactory instanceof SourceNodeFactory) {
+                node = new TopologyDescription.Source(
+                    nodeName,
+                    getSourceTopicListOrPatternAsString(nodeName));
+
+            } else if (nodeFactory instanceof ProcessorNodeFactory) {
+                node = new TopologyDescription.Processor(
+                    nodeName,
+                    getStoresOfNode(nodeName));
+
+            } else if (nodeFactory instanceof SinkNodeFactory) {
+                node = new TopologyDescription.Sink(
+                    nodeName,
+                    nodeToSinkTopic.get(nodeName));
+
+            } else {
+                throw new TopologyBuilderException("Unknown node type: " + nodeFactory.getClass().getName());
+            }
+
+            nodesByName.put(nodeName, node);
+        }
+
+        return nodesByName;
+    }
+
+    private String getSourceTopicListOrPatternAsString(final String sourceName) {
+        final Pattern sourcePattern = nodeToSourcePatterns.get(sourceName);
+
+        if (sourcePattern != null) {
+            return sourcePattern.pattern();
+        } else {
+            String sourceTopics = nodeToSourceTopics.get(sourceName).toString();
+            sourceTopics = sourceTopics.substring(1, sourceTopics.length() - 1); // trim first and last, ie. []
+
+            return sourceTopics;
+        }
+    }
+
+    private Set<String> getStoresOfNode(final String processorNode) {
+        final Set<String> stores = new HashSet<>();
+
+        for (final Map.Entry<String, StateStoreFactory> store : stateFactories.entrySet()) {
+            if (store.getValue().users.contains(processorNode)) {
+                stores.add(store.getKey());
+            }
+        }
+
+        return stores;
+    }
+
+    private void connectAllNodesToTheirPredecessorsAndSuccessors(final HashMap<String, TopologyDescription.Node> nodesByName) {
+        for (final TopologyDescription.Node node : nodesByName.values()) {
+            final NodeFactory nodeFactory = nodeFactories.get(node.name());
+
+            final String[] predecessorNames;
+            if (nodeFactory instanceof ProcessorNodeFactory) {
+                predecessorNames = ((ProcessorNodeFactory) nodeFactory).parents;
+
+            } else if (nodeFactory instanceof SinkNodeFactory) {
+                predecessorNames = ((SinkNodeFactory) nodeFactory).parents;
+
+            } else {
+                assert nodeFactory instanceof SourceNodeFactory;
```

Comment:
> it is not going to be anything else
That's true atm. But why do we write a test, the code work already :) (same miss leading argument) -> it's a guard against adding new "types" of `NodeFactory` -- it forces you to update this part of the code, too (otherwise you might miss it) And yes, it's private, thus it's an assert and no exception.

Not sure who the interface change resolve this issue? A `SourceNodeFactory` will have have `#parents` and thus, we will need to cast here... Or do I miss something?